### PR TITLE
Plasmapy (second attempt)

### PR DESCRIPTION
**Pull request by StanczakDominik**
Hopefully supersedes #4793 . I made another PR/fork so as to not pollute the earlier one/branch (I tend to be git messy when working with these), and to more easily test whether the current setup I have is enough to fix the previous issues without making a PyPI release prematurely.

### CODE REVIEW

**Comment by StanczakDominik** [comment]
Well, it looks like this is kind of building! I would very much  appreciate feedback on all the horrible things I'm probably doing here without knowing about it :D

**Comment by StanczakDominik** [comment]
Todo:

* [ ] move from test release on PyPI to factual release into github PlasmaPy tree (I didn't merge into main 0.1 branch because I wanted a release that built on conda first, and I was not sure whether the one I had would do)

File: recipes/plasmapy/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    - plasmapy.utils
+    - plasmapy.utils.tests
+    - plasmapy.visualization
+    - plasmapy.visualization.tests
+
+about:
+  home: https://github.com/PlasmaPy/PlasmaPy
+  license: BSD-3-Clause
+  license_family: BSD
+  #license_file: LICENSE.md
```
Comment (marcelotrevisani) [inline recipes/plasmapy/meta.yaml]: If the ``license`` is not being packaged you can pack it along with the recipe. After that open a PR to include the license

File: recipes/plasmapy/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  summary: Python package for plasma physics
+  description: 'PlasmaPy is a community-developed and community-driven Python package for plasma physics.'
+  doc_url: 'http://plasmapy.readthedocs.io/en/docs/'
+  dev_url: 'https://github.com/plasmapy/plasmapy'
+
+extra:
+  recipe-maintainers:
+    - SolarDrew
```
Comment (marcelotrevisani) [inline recipes/plasmapy/meta.yaml]: @SolarDrew , do you agree to be a maintainer as well?

File: recipes/plasmapy/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  host:
+    - python
+    - cython
+    - setuptools
+    - numpy
+    - scipy
+    - astropy
+  run:
+    - python
+    - numpy
```
Comment (marcelotrevisani) [inline recipes/plasmapy/meta.yaml]: are you compiling against ``numpy`` ?

File: recipes/plasmapy/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    - scipy
+    - astropy
+    - pytest
+    - mpmath
+    - h5py
+    - lmfit
+    - matplotlib
+    - colorama
+    - pytest-astropy
```
Comment (marcelotrevisani) [inline recipes/plasmapy/meta.yaml]: Do you need this pytest plugin to run this project?

File: recipes/plasmapy/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  version: {{ version }}
+
+source:
+  fn: plasmapy-{{version}}.tar.gz
+  url: https://pypi.io/packages/source/p/plasmapy/plasmapy-{{ version }}.tar.gz
+  sha256: 2a2bbc064a8b5726bc32447973babba9355e5c7d9483b1eebc82c6e83a9a173e
+
+build:
+  number: 0
+  script: python setup.py install --offline --single-version-externally-managed --record=record.txt
```
Comment (marcelotrevisani) [inline recipes/plasmapy/meta.yaml]: Please, change that to:
```
script: "{{ PYTHON }} -m pip install . --no-deps --ignore-installed --no-cache-dir -vvv"
```

Comment (StanczakDominik) [reply recipes/plasmapy/meta.yaml]: The original PR (#4793) did it this way (see https://github.com/conda-forge/sunpy-feedstock/blob/master/recipe/meta.yaml#L14 ), and @SolarDrew claimed needing to do it this way in the previous PR, seemingly following SunPy's example. This seemed to make sense to me, as both are taking advantage of Astropy's package ecosystem. It seems likely that this change caused the most recent build failure.

File: recipes/plasmapy/meta.yaml (hunk 4 above)
Comment (StanczakDominik) [reply recipes/plasmapy/meta.yaml]: When I was testing this earlier, it seemed to crash without the plugin. Also, see SunPy's recipe: https://github.com/conda-forge/sunpy-feedstock/blob/master/recipe/meta.yaml#L46

File: recipes/plasmapy/meta.yaml (hunk 5 above)
Comment (StanczakDominik) [reply recipes/plasmapy/meta.yaml]: Confirmed by a6942ed

File: recipes/plasmapy/meta.yaml (hunk 3 above)
Comment (StanczakDominik) [reply recipes/plasmapy/meta.yaml]: Well, we have a dependency on astropy, and that has a hard dependency on numpy... Also, we have a some Cython scripts that are going to rely on Numpy.

Also, it works for SunPy!

File: recipes/plasmapy/meta.yaml (hunk 1 above)
Comment (StanczakDominik) [reply recipes/plasmapy/meta.yaml]: I added it directly in 
9980325 .

File: recipes/plasmapy/meta.yaml (hunk 5 above)
Comment (marcelotrevisani) [reply recipes/plasmapy/meta.yaml]: This is the new pattern in ``conda-forge``, that's why I requested that

File: recipes/plasmapy/meta.yaml (hunk 2 above)
Comment (SolarDrew) [reply recipes/plasmapy/meta.yaml]: Sorry, just catching up on this. Yes, I'm happy to be a maintainer :)

**Comment by StanczakDominik** [comment]
Nah, I honestly have no idea how to do this with pip install. I thought it may have been the no-deps thing, but doesn't seem likely anymore due to b4d0208 failing on some conda forge here. Can we please just skip that and go with setup.py? I'm sorry this doesn't exactly conform to the new standard, but I'm at a complete loss here.

**Comment by StanczakDominik** [comment]
Given that we're going to be releasing PlasmaPy 0.2 soon-ish ( https://github.com/PlasmaPy/PlasmaPy/issues/566 ), this can probably wait until then.

**Comment by StanczakDominik** [comment]
@conda-forge-admin, please rerender

**Comment by StanczakDominik** [comment]
https://www.youtube.com/watch?v=AXwGVXD7qEQ

**Comment by StanczakDominik** [comment]
@conda-forge/staged-recipes it's finally working! :)

**Comment by StanczakDominik** [comment]
I think I might have messed up the mention the last time.

I call upon thee, @conda-forge/staged-recipes !

Since that doesn't seem to work, ping @marcelotrevisani - apologies, since I see you're busy, but I'm not sure where else to go with this :(

File: recipes/plasmapy/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  skip: True    # [py < 36]
+
+requirements:
+  build:
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+  host:
+    - python 3.6.*
```
Comment (mariusvniekerk) [inline recipes/plasmapy/meta.yaml]: Is it possible for these pins to float and be driven by the general set of pins for conda-forge.

In particular we are no longer setting our lower bound to versions of scipy and numpy that are this old.  Also python 3.7 is pretty solid by now

Comment (StanczakDominik) [reply recipes/plasmapy/meta.yaml]: It took me a while to understand what you mean (woo 6 AM), but I think I get it now. Is commit 187db7cef042c18add25f5fb59284e8792f3689a all right?

File: recipes/plasmapy/meta.yaml [hunk 7]
```diff
@@ -19,13 +19,13 @@ requirements:
     - {{ compiler("c") }}
     - {{ compiler("cxx") }}
   host:
-    - python >= 3.6
-    - cython >= 0.28
+    - python
```
Comment (StanczakDominik) [inline recipes/plasmapy/meta.yaml]: PlasmaPy requires 3.6 as we have fstrings basically everywhere. Could that potentially become a problem?

Comment (chrisburr) [reply recipes/plasmapy/meta.yaml]: You should only use this method of constraining the version for noarch packages. As this doesn’t have noarch: python, the constraint comes from the skip: True [py<36].

Also, version constraints should be formatted with one space before the operator and none after.

File: recipes/plasmapy/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  number: 0
+  script: python setup.py install --offline --single-version-externally-managed --record=record.txt
+  skip: True    # [py < 36]
+
+requirements:
+  build:
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+  host:
+    - python >= 3.6
```
[the remark this answers is not in the document]
Comment (StanczakDominik) [reply recipes/plasmapy/meta.yaml]: Ah, that clears it up - thanks! I had forgotten the `skip` command was there.

File: recipes/plasmapy/meta.yaml (hunk 7 above)
Comment (StanczakDominik) [reply recipes/plasmapy/meta.yaml]: Thanks, I'm learning a lot here! :)

**Comment by StanczakDominik** [comment]
I know what the problem is! Same thing as before in https://github.com/conda-forge/staged-recipes/pull/6694#issuecomment-427025131 - it's astropy-helpers trying to update itself. (please see https://github.com/astropy/conda-channel-astropy/blob/401bc10bde9114ed3a9fea33332eae638e19622b/requirements.yml#L22-L26 and the recently released SunPy 1.0 - https://github.com/conda-forge/sunpy-feedstock/blob/master/recipe/meta.yaml#L14 ).

I think the only solution right now, except for us getting off astropy-helpers (might take a while...) is reverting 47c25f1a5a9696ab4efc66d0ba711a37b758be9f, using setup.py and passing the `--offline` flag.

**Comment by chrisburr** [comment]
I think you can use a similar patch to this PR: https://github.com/conda-forge/staged-recipes/pull/8490/files

**Comment by StanczakDominik** [comment]
That patch solution is darn clever - I definitely wouldn't have come up with it. I'll still test `noarch`.

**Comment by chrisburr** [comment]
Assuming this has compiled C or C++ it can’t be made noarch. If it doesn’t contain compiled code I’ll show you how to do things a little differently.

**Comment by StanczakDominik** [comment]
Perhaps let's skip the noarch thing for now. I'm honestly considering replacing the cython we have in PlasmaPy with Numba anyway - I'm not 100% sure we'll keep it.

File: recipes/plasmapy/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    - {{ compiler("cxx") }}
+  host:
+    - python
+    - pip
+    - cython >=0.28
+    - setuptools
+    - numpy >=1.13
+    - scipy >=0.19
+    - astropy >=2.0
+    - colorama >=0.3
```
Comment (chrisburr) [inline recipes/plasmapy/meta.yaml]: I suspect you don't need `colorama` and `scipy` in `host`.

(`host` is only for dependencies which are needed to run the `setup.py` script)

Comment (StanczakDominik) [reply recipes/plasmapy/meta.yaml]: I suspect we wouldn't necessarily need them, but right now they're in `install_requires`, see here: https://github.com/PlasmaPy/PlasmaPy/blob/master/setup.cfg#L12

It's likely overkill as we're [still figuring out the development process (pic related)](https://hips.hearstapps.com/ghk.h-cdn.co/assets/cm/15/11/54ffe5266025c-dog1.jpg?crop=1xw:1.0xh;center,top&resize=5760:*), but we released 0.2.0 with those in install_requires. We'll probably move them to optional in the future, however, for now I'm betting that this test is going to blow up.

Comment (StanczakDominik) [reply recipes/plasmapy/meta.yaml]: I completely agree that it's probably a bad idea to have them there.

Comment (chrisburr) [reply recipes/plasmapy/meta.yaml]: IIRC `setup_requires` corresponds to `host` dependencies and `install_requires` corresponds to `run` dependencies. It clearly failed though so it doesn't really matter.

I'll merge this once it passes (ping me if I forget :wink:).

Comment (StanczakDominik) [reply recipes/plasmapy/meta.yaml]: @chrisburr all green across the board! :smile:

**Comment by StanczakDominik** [comment]
Awesome! Much thanks to everyone involved! :)

**Pull request merged**